Introduced future finite element indices.

marcfehling opened the pull request:
@davydden, @bangerth and I have developed a generalized interface for the new `p-adaptivity` functionality that is way more versatile than the current one: Instead of setting flags, the user is able to specify the finite elements that a cell will have after the triangulation changes by setting a `future_fe_index`.

---

Removed funtions:
```
DoFCellAccessor::p_refine_flag_set()
DoFCellAccessor::set_p_refine_flag()
DoFCellAccessor::clear_p_refine_flag()

DoFCellAccessor::p_coarsen_flag_set()
DoFCellAccessor::set_p_coarsen_flag()
DoFCellAccessor::clear_p_coarsen_flag()

DoFCellAccessor::active_fe_index_after_p_refinement_and_coarsening()
```
New functions:
```
DoFCellAccessor::future_fe_index()
DoFCellAccessor::set_future_fe_index()
DoFCellAccessor::future_fe_index_set()
DoFCellAccessor::clear_future_fe_index()
```

---

Please make sure that this will be part of the `Release 9.1` since major parts of the `p-adaptivity` interface are affected. (We would have to deprecate the above functions otherwise, and I would like to avoid that.)
@bangerth -- This is the PR I was talking about.

Part of #7515 and #3511.

~~PS: The first commit is part of #7933. I will rebase once it got merged.~~ EDIT: Done.

davydden commented on include/deal.II/dofs/dof_accessor.h:
```diff
@@ -1970,82 +1955,110 @@ class DoFCellAccessor : public DoFAccessor<DoFHandlerType::dimension,
 
   /**
-   * Return whether the p-refinement flag is set or not.
+   * Return the fe_index of the future finite element that this object
```
Not sure my wording is better, but maybe something along the lines:

> Return the `fe_index` of the finite element that will be used in cell next time triangulation is changed (i.e. due to refinement, coarsening or repartitioning). If the "future finite element" has not be specified via the set_future_fe_index() function for this cell, the current one will be used, in which case the fe_index of the active finite element will be returned.

davydden commented on include/deal.II/dofs/dof_accessor.h:
```diff
@@ -1970,82 +1955,110 @@ class DoFCellAccessor : public DoFAccessor<DoFHandlerType::dimension,
+   * Set the fe_index of the future finite element that this object
+   * will be assigned to after the triangulation changed (due to refinement
+   * and/or coarsening, or repartitioning). A previously assigned future finite
+   * element will be overwritten.
```
I would remove the same docu and instead write

> See future_fe_index() for more information.

bangerth replied:
I agree.

davydden commented on include/deal.II/dofs/dof_accessor.h:
```diff
@@ -1970,82 +1955,110 @@ class DoFCellAccessor : public DoFAccessor<DoFHandlerType::dimension,
    *
-   * @note The DoFHandler in use has to be of type hp::DoFHandler.
-   * An exception is thrown otherwise.
+   * @note Since degrees of freedom only exist on active cells for
```
ditto

davydden commented on include/deal.II/dofs/dof_accessor.h:
```diff
@@ -1970,82 +1955,110 @@ class DoFCellAccessor : public DoFAccessor<DoFHandlerType::dimension,
-
-  /**
-   * Clear the p-coarsening flag.
+   * @note Since degrees of freedom only exist on active cells for
```
ditto

marcfehling commented:
My idea was to utilize the future_fe_indices only when requested by the user. I chose these particular four functions because I had problems determining if such an index has been set or not. Consider the following cases:

1. If we would leave full control to the user, they would have to keep track of both active and future fe, which would be too tedious. They would have to set future_fe_indices, although they will probably never end up using p-adaptation.
```
dof_handler.set_fe(collection); // active_fe_index=0 future_fe_index=0
cell->set_active_fe_index(3);   // active_fe_index=3 future_fe_index=0
cell->set_future_fe_index(5);   // active_fe_index=3 future_fe_index=5
cell->set_active_fe_index(4);   // active_fe_index=4 future_fe_index=5
```

2. We could let `set_active_fe_index()` set both active and future index, but this will overwrite future indices that have been previously assigned, and would lead to ambiguous results.
```
dof_handler.set_fe(collection); // active_fe_index=0 future_fe_index=0
cell->set_active_fe_index(3);   // active_fe_index=3 future_fe_index=3
cell->set_future_fe_index(5);   // active_fe_index=3 future_fe_index=5
cell->set_active_fe_index(4);   // active_fe_index=4 future_fe_index=4 ?
```

3. We could let `set_active_fe_index()` set both active and future index only when synchronized, but this would again end in ambiguous results.
```
dof_handler.set_fe(collection); // active_fe_index=0 future_fe_index=0
cell->set_active_fe_index(3);   // active_fe_index=3 future_fe_index=3
cell->set_future_fe_index(5);   // active_fe_index=3 future_fe_index=5
cell->set_active_fe_index(5);   // active_fe_index=5 future_fe_index=5
cell->set_active_fe_index(4);   // active_fe_index=4 future_fe_index=4 ?
```

I chose this interface in order to exclude ambiguity. Do you have an idea how to simplify it?

marcfehling commented:
Further, with the current implementation, `future_fe_indices` can be kept temporary. In general, we don't even have to communicate them between processors.

However, looking at the `communicate_active_fe_indices()` functions, we indeed communicate the future indices as well, and I can not recall why. I guess I wanted to give the user the ability to keep their `future_fe_index` setup while repartitioning the mesh.

What do you think? Shall we invalidate all `future_fe_indices` upon repartitioning (when no refinement is involved), or shall we communicate them in this case as well?

marcfehling commented:
@davydden The function you had in mind can be made `const` since it does not change the `DoFCellAccessor` object itself, but internal structures of the `hp::DoFHandler` (similarly to the `active_fe_indices`).

bangerth commented on include/deal.II/dofs/dof_accessor.h:
```diff
@@ -1970,82 +1955,110 @@ class DoFCellAccessor : public DoFAccessor<DoFHandlerType::dimension,
+   * parallel::shared::Triangulation or parallel::distributed::Triangulation
+   * classes, it is only allowed to call this function on locally
+   * owned or ghost cells. No information is available on artificial cells.
+   * Furthermore, @p future_fe_index information is exchanged from locally
```
```suggestion
   * Furthermore, @p future_fe_index information is only exchanged from locally
```

bangerth commented on include/deal.II/dofs/dof_accessor.h:
```diff
@@ -1970,82 +1955,110 @@ class DoFCellAccessor : public DoFAccessor<DoFHandlerType::dimension,
+   * Furthermore, @p future_fe_index information is exchanged from locally
+   * owned cells on one processor to other processors where they may be
+   * ghost cells, during the call to hp::DoFHandler::set_fe() and
+   * hp::DoFHandler::distribute_dofs(). See the documentation of hp::DoFHandler
```
```suggestion
   * hp::DoFHandler::distribute_dofs(). If you call `set_future_fe_index()` on a cell after the
   * call to `distribute_dofs()`, then this information will not be propagated to other processors
   * who may have this cell as a ghost cell. See the documentation of hp::DoFHandler
```

bangerth commented on include/deal.II/hp/dof_level.h:
```diff
@@ -505,8 +607,7 @@ namespace internal
       ar & this->dof_offsets;
-      ar & this->p_refine_flags;
-      ar & this->p_coarsen_flags;
+      ar & this->future_fe_indices;
```
You asked in a previous comment whether it makes sense to serialize these flags. One can probably think about it in different ways, but it definitely doesn't hurt.

marcfehling replied:
I have no idea where that comment went :-)

I decided to keep it, since we do the same for `cell_refine|coarsen_flags` in `TriaLevel`.
https://github.com/dealii/dealii/blob/095d16fdaf8d016141b85f7c43a727192b20ea2b/include/deal.II/grid/tria_levels.h#L272

marcfehling commented:
Addressed your suggestions.

I introduced a function that checks whether all `future_fe_indices` have been cleared before repartitioning happens. For this, I introduced a signal that will be called in the corresponding `GridTools` functions. For the `p::d` case, a corresponding signal already exists to which I connected this check.

After my local testsuite passes, I will squash both commits to one. Feel free to have a final look over this PR :)

davydden commented:
/rebuild

bangerth commented on include/deal.II/grid/tria.h:
```diff
@@ -2206,10 +2206,18 @@ class Triangulation : public Subscriptor
      * This signal is triggered at the end of execution of the
      * Triangulation::execute_coarsening_and_refinement() function when the
-     * triangulation has reached its final state
+     * triangulation has reached its final state.
```
:-) Nice job!

bangerth commented:
/rebuild

kronbichler commented:
@marcfehling the serial build in the CI complains:
```
lib/libdeal_II.g.so.9.1.0-pre: error: undefined reference to 'dealii::internal::hp::DoFLevel::invalid_active_fe_index'
```
Would you mind checking this?

marcfehling commented:
For some reason our testers have an issue with this line:
```
private:
  static constexpr type var = value;
```
I changed it to:
```
private:
  static const type var;
```
and moved the definition to the `cpp` file. I hope that helped (it does at least work on my machine).

kronbichler commented:
> I hope that helped (it does at least work on my machine)

It has passed the point where it failed before.

masterleinad commented:
> For some reason our testers have an issue with this line:
> 
> ```
> private:
>   static constexpr type var = value;
> ```
> 
> I changed it to:
> 
> ```
> private:
>   static const type var;
> ```
> 
> and moved the definition to the `cpp` file. I hope that helped (it does at least work on my machine).

We have seen that problem in many other places before...
